src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -1304,4 +1313,79 @@ public override Task<ImplementationResult> ImplementationAsync(DelegatedPosition
     }
 
     private record struct DelegationRequestDetails(string LanguageServerName, Uri ProjectedUri, ITextBuffer TextBuffer);
+
+    private class TelemetryReportingLSPRequestInvoker : LSPRequestInvoker
+    {
+        private LSPRequestInvoker _requestInvoker;
+        private ITelemetryReporter _telemetryReporter;
+        public TelemetryReportingLSPRequestInvoker(LSPRequestInvoker requestInvoker, ITelemetryReporter telemetryReporter)
+        {
+            _requestInvoker = requestInvoker;
+            _telemetryReporter = telemetryReporter;
+        }
+
+        public override Task<IEnumerable<ReinvokeResponse<TOut>>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(string method, string contentType, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(method, contentType, parameters, cancellationToken);
+        }
+
+        public override Task<IEnumerable<ReinvokeResponse<TOut>>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(string method, string contentType, Func<JToken, bool> capabilitiesFilter, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(method, contentType, capabilitiesFilter, parameters, cancellationToken);
+        }
+
+        public override IAsyncEnumerable<ReinvocationResponse<TOut>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(ITextBuffer textBuffer, string method, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(textBuffer, method, parameters, cancellationToken);
+        }
+
+        public override IAsyncEnumerable<ReinvocationResponse<TOut>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(ITextBuffer textBuffer, string method, Func<JToken, bool> capabilitiesFilter, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(textBuffer, method, capabilitiesFilter, parameters, cancellationToken);
+        }
+
+        public override Task<ReinvokeResponse<TOut>> ReinvokeRequestOnServerAsync<TIn, TOut>(string method, string languageServerName, TIn parameters, CancellationToken cancellationToken)
+        {
+            using (_telemetryReporter?.BeginBlock(nameof(ReinvokeRequestOnServerAsync), Severity.Normal, ImmutableDictionary<string, object?>.Empty
+                    .Add("eventscope.method", method)
+                    .Add("eventscope.languageservername", languageServerName)
+            ))
```

Comment:
Personally I would extract this to a helper method, just to make this neater. ie make this line be something like `using (Track(nameof(ReinvokeRequestOnServerAsync), method, languageServerName))`

Reply:
👍